pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -161,21 +169,26 @@ type reconciler struct {
 	// idAllocator provides interfaces to allocate and release uint32 id.
 	idAllocator *idAllocator
 
-	// priorityAssigner provides interfaces to manage OF priorities.
-	priorityAssigner *priorityAssigner
-
-	// priorityMutex prevents concurrent priority re-assignments
-	priorityMutex sync.RWMutex
+	// priorityAssigners provides interfaces to manage OF priorities for each OVS table.
+	priorityAssigners map[binding.TableIDType]*tablePriorityAssigner
 }
 
 // newReconciler returns a new *reconciler.
 func newReconciler(ofClient openflow.Client, ifaceStore interfacestore.InterfaceStore) *reconciler {
+	cnpTables := append(openflow.GetCNPEgressTables(), openflow.GetCNPIngressTables()...)
```

Comment:
It seems a packet will need to go through 10 + 2 tables just for policy enforcement, and perhaps more tables when dynamic tiering is introduced? In each table, 3 clauses need to be computed, which means 36 times of hash and search.
Hope it wouldn't add much overhead. Could we check the impact by testing latency and TCP_CRR?

Reply:
Will do a performance check and post the result here.

Reply:
Some measurements of latency between the original OVS pipeline and new OVS pipeline. For the original pipeline, I populated table 45,50,85,90 with conjuctive flows that cover src_addr, dst_addr and port/protocol. For the new pipeline, 45-50 and 85-90 are all populated with such flows. The traffic being tested in not specified by any NetworkPolicy so it must go through the entire pipeline.
Latency results:
```
<Same host, after OVS pipeline changed to use additional tables>
--------------------------------- iperf ----------------------------------------
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  49.4 GBytes  42.4 Gbits/sec  1004             sender
[  5]   0.00-10.04  sec  49.4 GBytes  42.2 Gbits/sec                  receiver
--------------------------------- ping -----------------------------------------
50 packets transmitted, 50 received, 0% packet loss, time 57164ms
rtt min/avg/max/mdev = 0.054/0.118/0.408/0.114 ms
-------------------------------- TCP_CRR ---------------------------------------
/ # netperf -v 0 -P 0 -t TCP_CRR  -H 10.10.1.58 -- -r 256,256 -b 1 -D
1405.93

<Same host, before OVS pipeline changed to use additional tables>
--------------------------------- iperf ----------------------------------------
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  49.4 GBytes  42.4 Gbits/sec  1794             sender
[  5]   0.00-10.04  sec  49.4 GBytes  42.2 Gbits/sec                  receiver
--------------------------------- ping -----------------------------------------
50 packets transmitted, 50 received, 0% packet loss, time 50191ms
rtt min/avg/max/mdev = 0.051/0.074/0.567/0.077 ms
-------------------------------- TCP_CRR ---------------------------------------
/ # netperf -v 0 -P 0 -t TCP_CRR  -H 10.10.1.58 -- -r 256,256 -b 1 -D
7409.66
```
Admittedly the TCP_CRR is influenced quite a bit by the new tables. RTT and throughput seem to be less affected. The other alternative to this design could be to use a single table for Tiers other than the default one but I'm if that's the path we want to go down for this release @jianjuns @abhiraut

Reply:
seems like a decent drop in rate because of the additional overhead.. with dynamic tiering.. users with more tiers may face such a situation .. but most users may have 3-5 tiers (if they do tiering) .. we should document this overhead.. 
on the other hand.. i think we were also exploring of sharing tables for tiers and having one table dedicated for the default tier which will eventually be the most populated tier..

Reply:
It's expected that throughput (iperf) is not affected because only the first packet would have the overhead.
The latency might be worse, your test used only 1 ICMP session so only the first packet had the overhead. You may try creating 50 ICMP sessions, each of which has only 1 request to see the real impact on latency, like
```
for i in `seq 50`; do ping -c 1 X.X.X.X; done
```

The impact on TCP_CRR also means it would consume more CPU cycles for container traffic. I understand it's quite tight to change the design at the moment, but could we make sure that the packets won't go the extra tables when ClusterNetworkPolicy is not enabled so clusters with only core features enabled can still perform good.

For the design itself, I'm thinking given the limit of priorities is 65K and normally a Node has up to hundreds of Pods, one table should be enough for most cases as it can support hundreds of rules applied to each Pod. Could we design the dynamic priority assignment with tiers too? So that if the number of rules enforced on one Node is less than 65K, it will consume a single table, and it could extend one or more tables only when it beyonds 65K rules, which is unlikely to happen.

Reply:
Let me make the change for bypassing the cnptables when the feature is not enabled. As for the design, the priority assigner already supports sorting priorities from different tiers, so theoretically putting all tiers in a single table can be done with not much change involved (the only part should be InitialOFPriority function). I remember the tier per table design was to answer to scale demands, and dynamically reassign flows from one table to another is a hustle in runtime. @abhiraut given the timeframe I can try to pull the second design off today, which is to have a table dedicated for default tier and share a table for all other tiers.

Reply:
Sorry, what is the conclusion on single table or multi-tables? I assume we still keep one table per tier for this release, and optimize later?

Another question - if no NP in a tier on the local host, it should not be overhead to have a table right, as there is just a default flow to jump to another table?

Reply:
To answer the second question, yes there should not be overhead of no rules in a tier applies to pod on that node. This is also why in this PR I'm keeping the one tier per table design, and simply bypassing these tables if CNP is not enabled. I will open another PR which tries to put the other 4 tier in one table, and see if it can be merged in time for v0.9.

Reply:
Ok, I think fine to keep the current implementation in 0.9, and we can decide later what to do when we really have some customers of tiered policies.